javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -266,46 +283,47 @@ private void setCurrentArray() {
     BitMappedTrie<T> filter(Predicate<? super T> predicate) {
         Objects.requireNonNull(predicate, "predicate is null");
 
-        final Object[] results = new Object[length()];
+        final Object results = type.newInstance(length());
         final int length = this.<T> visit((index, leaf, start, end) -> {
             for (int i = start; i < end; i++) {
-                final T value = leaf[i];
+                final T value = type.getAt(leaf, i);
                 if (predicate.test(value)) {
-                    results[index++] = value;
+                    type.setAt(results, index++, value);
                 }
             }
             return index;
         });
 
         return (length() == length)
                ? this
-               : BitMappedTrie.ofAll(copyRange(results, 0, length));
+               : BitMappedTrie.ofAll(type.copyRange(results, 0, length));
     }
 
     <U> BitMappedTrie<U> map(Function<? super T, ? extends U> mapper) {
         Objects.requireNonNull(mapper, "mapper is null");
 
-        final Object[] results = newInstance(length);
+        final Object results = obj().newInstance(length);
         this.<T> visit((index, leaf, start, end) -> {
             for (int i = start; i < end; i++) {
-                results[index++] = mapper.apply(leaf[i]);
+                obj().setAt(results, index++, mapper.apply(type.getAt(leaf, i)));
             }
             return index;
         });
 
         return BitMappedTrie.ofAll(results);
     }
 
-    int length() { return length; }
+    int length()                         { return length; }
+    private boolean cannotAdd(T element) { return type != obj() && ((element == null) || (type.type() != asPrimitive(element.getClass()))); }
```

Comment:
> We don't need to check the types. Because we already use T

you can add a `String` to a primitive array via `narrow`, e.g.
```java
@Test
public void shouldNarrowPrimitives() {
    final Vector<Object> object = Vector.narrow(range(0, 2));
    Vector<Object> actual = object.append("String");
    assertThat(actual).isEqualTo(of(0, 1, "String"));
}
```

(added this test explicitly now. `shouldBehaveLikeArray` already tested this implicitly)

> arrayType.isPrimitive() && (element == null || !arrayType.isPrimitive());

`A & (B | !A)` == `A && B`, since `!A` is always `false` (typo?)

---

let me know if you still prefer to change this